### PR TITLE
[Feature](bangc-ops): add concat binary operator.

### PR DESCRIPTION
**Pull request by zhengleiZL**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

将concat算子以二进制形式合入mluops

## 2. Modification

增加concat算子

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff3
    - [x] mlu diff3 == 0
    - [x] mlu diff3_1 == 0
    - [x] mlu diff3_2 == 0

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
[----------] Global test environment tear-down
[ SUMMARY  ] Total 131 cases of 79 op(s).
ALL PASSED.
[==========] 131 test cases from 79 test suites ran. (30884 ms total)
[  PASSED  ] 131 test cases.

```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ * Number of tensors needed to be concatenated.
+ * @param[in] axis
+ * Dimension along which to be concatenated. The value must be in the range of [-rank, rank),
+ * where rank is the number of dimensions in the input tensors,
+ * and negative \b axis refers to ``axis + rank``.
+ * @param[in] inputs_desc
+ * The list of descriptors of input tensors. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] inputs
+ * A host pointer to a list of MLU pointers, which point to the MLU memory that store the
```
Comment (Unireverse) [inline bangc-ops/mlu_op.h:15313]:
```suggestion
 * A host pointer to a list of MLU pointers, which points to the MLU memory that stores the
```

**Review by Unireverse — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+            const mluOpTensorDescriptor_t output_desc,
+            void *output);
+
+// Group:Concat
+/*!
+ * @brief Returns in \b size the size of the MLU memory that is used as an extra workspace to
+ * optimize the concat operation.
+ *
+ * @param[in] handle
+ * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15402]:
```suggestion
 * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues.
```

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+            const mluOpTensorDescriptor_t inputs_desc[],
+            const void *const inputs[],
+            void *workspace,
+            size_t workspace_size,
+            const mluOpTensorDescriptor_t output_desc,
+            void *output);
+
+// Group:Concat
+/*!
+ * @brief Returns in \b size the size of the MLU memory that is used as an extra workspace to
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15398]:
```suggestion
 * @brief Returns the size of the MLU memory used as an extra workspace to
```

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ * optimize the concat operation.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15399]:
```suggestion
 * optimize ::mluOpConcat.
```

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ *
+ * @par Scale Limitation
+ * - The parameters must meet the following requirements:
+ *   - The parameter \b concat_num should be greater than 0.
+ *   - The number of dimensions of all tensors must match, including inputs and output.
+ *   - All dimensions except \b axis must be equal, and the dimension of output on \b axis must be
+ *     equal to the sum of input dimensions on \b axis.
+ *
+ * @par API Dependency
+ * - Before calling this function to implement concat, you need to call
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15356]:
```suggestion
 * - Before calling this function to implement concatenation, you need to call
```

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ *   ::mluOpGetConcatWorkspaceSize to get the extra space size needed in concat operation.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15357]:
```suggestion
 *   ::mluOpGetConcatWorkspaceSize to get the required extra space size.
```

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ * Pointer to the MLU memory that stores the output tensor.
+ *
+ * @par Return
+ * -  ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM, ::MLUOP_STATUS_ALLOC_FAILED
+ *
+ * @par Formula
+ * - See "Concat Operator" section in "Cambricon MLUOP User Guide" for details.
+ *
+ * @par Data Type
+ * - The (I/O)function supports the following byte-width data types for \b input and \b output tensors.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15336]:
```suggestion
 * - The I/O function supports the following byte-width data types for \b input and \b output tensors.
```

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ *   The byte width of a data type can be obtained by ::mluOpGetSizeOfDataType function.
+ *   <b>Note that all the tensors must have the same data type. If the tensors are in
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15338]: 建议生成看下，这个格式 pdf中显示可能有问题

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] inputs
+ * A host pointer to a list of MLU pointers, which point to the MLU memory that store the
+ * input tensors.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the concat operation.
+ * For more information about workspace, see "Cambricon BANG C OPS User Guide". Because ::mluOpConcat
+ * does not need extra workspace, the \b workspace can be set to NULL.
+ * @param[in] workspace_size
+ * The size of the extra workspace in bytes that needs to be used in the concat operation.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15320]:
```suggestion
 * The size of the extra workspace in bytes.
```

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ * @param[in] inputs_desc
+ * The list of descriptors of input tensors. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] inputs
+ * A host pointer to a list of MLU pointers, which point to the MLU memory that store the
+ * input tensors.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the concat operation.
+ * For more information about workspace, see "Cambricon BANG C OPS User Guide". Because ::mluOpConcat
+ * does not need extra workspace, the \b workspace can be set to NULL.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15318]:
```suggestion
 * does not need extra workspace,  \b workspace can be set to NULL.
```

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ * where rank is the number of dimensions in the input tensors,
+ * and negative \b axis refers to ``axis + rank``.
+ * @param[in] inputs_desc
+ * The list of descriptors of input tensors. For detailed information,
+ * see ::mluOpTensorDescriptor_t.
+ * @param[in] inputs
+ * A host pointer to a list of MLU pointers, which point to the MLU memory that store the
+ * input tensors.
+ * @param[in] workspace
+ * Pointer to the MLU memory that is used as an extra workspace for the concat operation.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15316]:
```suggestion
 * Pointer to the MLU memory that is used as an extra workspace.
```

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
                   const mluOpTensorDescriptor_t output_desc,
                   void *output);
 
+// Group:Concat
+/*!
+ * @brief Concatenates the list of input tensors \b inputs along the given dimension \b axis.
+ *
+ * @param[in] handle
+ * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15301]:
```suggestion
 * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues.
```

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -15293,6 +15293,156 @@ mluOpStridedSlice(mluOpHandle_t handle,
...
+ * ::mluOpConcat operation. For detailed information, see ::mluOpHandle_t.
```
Comment (AquamarineSu) [inline bangc-ops/mlu_op.h:15302]:
```suggestion
 * For detailed information, see ::mluOpHandle_t.
```

**Pull request closed without merging**